### PR TITLE
remove adjust_logits_during_generation method

### PR DESCRIPTION
**Pull request by patil-suraj**
# What does this PR do?

This PR is the first split of #9811.

This PR
1.  introduces two new `generate`  and `config` arguments  and `LogitsProcessor`
    - `forced_bos_token_id` and `forced_eos_token_id`,  to force a specific start and end token. This is particularly useful for many to many and one to many translation models, so we can pass different language tokens as `forced_bos_token_id` to `generate`, 
    - `ForcedBOSTokenLogitsProcessor` and `ForcedEOSTokenLogitsProcessor`
2. Remove `adjust_logits_during_generation` method from all models (except `Marian`) and handle that use case using the newly introduced logits processors.
4. remove the `force_bos_token_to_be_generated` argument from `BartConfig`

For `Marian` we still need to keep the `adjust_logits_during_generation` method to force the model to not generate pad token. Adding the pad token to `bad_words_ids` does not resolve this issue, the score of `pad_token_id` needs to be set to `-inf` before calling `log_softmax`

### CODE REVIEW

File: src/transformers/generation_utils.py [hunk 1]
```diff
@@ -1866,7 +1893,7 @@ def beam_sample(
             )
             next_token_logits = outputs.logits[:, -1, :]
 
-            # adjust token scores (a no-op by default)
+            # hack: adjust tokens for Marian. For Marian we have to make sure that the `pad_token_id` cannot be generated both before and after the `F.log_softmax` operation.
```
Comment (patrickvonplaten) [inline src/transformers/generation_utils.py:1896]: (nit) maybe split into 2 lines

File: src/transformers/generation_utils.py [hunk 2]
```diff
@@ -1611,7 +1638,7 @@ def beam_search(
             )
             next_token_logits = outputs.logits[:, -1, :]
 
-            # adjust tokens for Bart, *e.g.*
+            # hack: adjust tokens for Marian. For Marian we have to make sure that the `pad_token_id` cannot be generated both before and after the `F.log_softmax` operation.
```
Comment (patrickvonplaten) [inline src/transformers/generation_utils.py:1641]: (nit) maybe split into 2 lines

File: src/transformers/generation_utils.py [hunk 3]
```diff
@@ -2150,7 +2177,7 @@ def group_beam_search(
                 # select outputs of beams of current group only
                 next_token_logits = outputs.logits[batch_group_indices, -1, :]
 
-                # adjust tokens for Bart, *e.g.*
+                # hack: adjust tokens for Marian. For Marian we have to make sure that the `pad_token_id` cannot be generated both before and after the `F.log_softmax` operation.
```
Comment (patrickvonplaten) [inline src/transformers/generation_utils.py:2180]: (nit) maybe split into 2 lines

File: src/transformers/models/rag/configuration_rag.py [hunk 4]
```diff
@@ -161,6 +166,9 @@ def __init__(
 
         self.use_cache = use_cache
 
+        if self.forced_eos_token_id is None:
```
Comment (patrickvonplaten) [inline src/transformers/models/rag/configuration_rag.py:169]: Nice I like it!

File: src/transformers/models/rag/modeling_rag.py [hunk 5]
```diff
@@ -1498,7 +1504,10 @@ def extend_enc_output(tensor, num_beams=None):
             encoder_input_ids=context_input_ids,
             bad_words_ids=bad_words_ids,
             min_length=min_length,
+            max_length=max_length,
             eos_token_id=eos_token_id,
+            forced_bos_token_id=forced_bos_token_id,
+            forced_eos_token_id=forced_eos_token_id,
```
Comment (patrickvonplaten) [inline src/transformers/models/rag/modeling_rag.py:1510]: Nice

**Review by patrickvonplaten — APPROVED** [review]
Great job! Really happy that we are able to get rid of the `adjust_logits_during_generation` method (except for Marian).

Just to be sure, we did check that the slow tests both in PT and TF for:

Bart, MBart, Pegasus, Marian, Blenderbot, BlenderbotSmall & RAG (only PT) pass correct?

**Comment by patil-suraj** [comment]
All slow tests passing in PT!

**Comment by patil-suraj** [comment]
@patrickvonplaten  (Bart, MBart, Pegasus, Marian, Blenderbot, BlenderbotSmall) slow tests are passing in TF as well.

File: src/transformers/configuration_utils.py [hunk 6]
```diff
@@ -131,6 +131,12 @@ class PretrainedConfig(object):
           logits when used for generation
         - **return_dict_in_generate** (:obj:`bool`, `optional`, defaults to :obj:`False`) -- Whether the model should
           return a :class:`~transformers.file_utils.ModelOutput` instead of a :obj:`torch.LongTensor`
+        - **forced_bos_token_id** (:obj:`int`, `optional`) -- The id of the token to force as the first generated token
+          after the :obj:`decoder_start_token_id`. Useful for multilingual models like `mBART
+          <https://huggingface.co/transformers/model_doc/mbart.html>`__ where the first generated token needs to be the
```
Comment (sgugger) [inline src/transformers/configuration_utils.py:136]:
```suggestion
          after the :obj:`decoder_start_token_id`. Useful for multilingual models like :doc:`mBART
          <../model_doc/mbart>` where the first generated token needs to be the
```
To have the link work on every version of the doc and stay in said version.

File: src/transformers/generation_logits_process.py [hunk 7]
```diff
@@ -520,3 +520,49 @@ def __call__(
...
+        if cur_len == 1:
+            num_tokens = scores.shape[1]
+            scores[:, [i for i in range(num_tokens) if i != self.bos_token_id]] = -float("inf")
+            scores[:, self.bos_token_id] = 0
+        return scores
+
+
+class ForcedEOSTokenLogitsProcessor(LogitsProcessor):
+    r"""
+    :class:`transformers.LogitsProcessor` that enforces the specified token as the last generated token when
```
Comment (sgugger) [inline src/transformers/generation_logits_process.py:548]:
```suggestion
    :class:`~transformers.LogitsProcessor` that enforces the specified token as the last generated token when
```

File: src/transformers/generation_logits_process.py [hunk 8]
```diff
@@ -520,3 +520,49 @@ def __call__(
...
+    `max_length` is reached.
```
Comment (sgugger) [inline src/transformers/generation_logits_process.py:549]:
```suggestion
    :obj:`max_length` is reached.
```

File: src/transformers/generation_logits_process.py [hunk 9]
```diff
@@ -520,3 +520,49 @@ def __call__(
...
+
+    Args:
+        max_length (:obj:`int`):
+            The maximum length of the sequence to be generated.
+        eos_token_id (:obj:`int`):
+            The id of the token to force as the last generated token when `max_length` is reached.
```
Comment (sgugger) [inline src/transformers/generation_logits_process.py:555]:
```suggestion
            The id of the token to force as the last generated token when :obj:`max_length` is reached.
```

File: src/transformers/generation_tf_utils.py [hunk 10]
```diff
@@ -137,6 +139,13 @@ def generate(
             use_cache: (:obj:`bool`, `optional`, defaults to :obj:`True`):
                 Whether or not the model should use the past last key/values attentions (if applicable to the model) to
                 speed up decoding.
+            forced_bos_token_id (:obj:`int`, `optional`):
+                The id of the token to force as the first generated token after the :obj:`decoder_start_token_id`.
+                Useful for multilingual models like `mBART
+                <https://huggingface.co/transformers/model_doc/mbart.html>`__ where the first generated token needs to
```
Comment (sgugger) [inline src/transformers/generation_tf_utils.py:145]:
```suggestion
                Useful for multilingual models like :doc:`mBART
                <../model_doc/mbart>` where the first generated token needs to
```

File: src/transformers/generation_utils.py [hunk 11]
```diff
@@ -720,6 +737,13 @@ def generate(
                 Whether or not to return the prediction scores. See ``scores`` under returned tensors for more details.
             return_dict_in_generate (:obj:`bool`, `optional`, defaults to `False`):
                 Whether or not to return a :class:`~transformers.file_utils.ModelOutput` instead of a plain tuple.
+            forced_bos_token_id (:obj:`int`, `optional`):
+                The id of the token to force as the first generated token after the :obj:`decoder_start_token_id`.
+                Useful for multilingual models like `mBART
+                <https://huggingface.co/transformers/model_doc/mbart.html>`__ where the first generated token needs to
```
Comment (sgugger) [inline src/transformers/generation_utils.py:743]:
```suggestion
                Useful for multilingual models like :doc:`mBART
                <../model_doc/mbart>` where the first generated token needs to
```

File: src/transformers/models/rag/modeling_rag.py [hunk 12]
```diff
@@ -1403,6 +1402,13 @@ def generate(
                 conditioned on the previously generated tokens :obj:`inputs_ids` and the batch ID :obj:`batch_id`. This
                 argument is useful for constrained generation conditioned on the prefix, as described in
                 `Autoregressive Entity Retrieval <https://arxiv.org/abs/2010.00904>`__.
+            forced_bos_token_id (:obj:`int`, `optional`):
+                The id of the token to force as the first generated token after the :obj:`decoder_start_token_id`.
+                Useful for multilingual models like `mBART
+                <https://huggingface.co/transformers/model_doc/mbart.html>`__ where the first generated token needs to
```
Comment (sgugger) [inline src/transformers/models/rag/modeling_rag.py:1408]:
```suggestion
                Useful for multilingual models like :doc:`mBART
                <mbart>` where the first generated token needs to
```

**Review by sgugger — APPROVED** [review]
Thanks for splitting this part in a new PR!

**Comment by patil-suraj** [comment]
Applied Sylvain's suggestions. Merging!

File: src/transformers/generation_logits_process.py [hunk 13]
```diff
@@ -520,3 +520,49 @@ def __call__(
...
+        bos_token_id (:obj:`int`):
+            The id of the token to force as the first generated token.
+    """
+
+    def __init__(self, bos_token_id: int):
+        self.bos_token_id = bos_token_id
+
+    def __call__(self, input_ids: torch.LongTensor, scores: torch.FloatTensor) -> torch.FloatTensor:
+        cur_len = input_ids.shape[-1]
+        if cur_len == 1:
```
Comment (lucadiliello) [inline src/transformers/generation_logits_process.py:539]: I do not get why it is not `if cur_len == 0:`

Comment (patrickvonplaten) [reply src/transformers/generation_logits_process.py:539]: Because the first token is always the `decoder_start_token_id` which is a dummy token.
